phoenix-core/src/test/java/org/apache/phoenix/index/IndexToolSingleRowVerifyTest.java

```diff
@@ -0,0 +1,634 @@
+            PreparedStatement ps =
+                    conn.prepareStatement(
+                            String.format(partialRowUpsert1, dataTableFullName, FIRST_ID, SECOND_ID,
+                                    FIRST_VALUE));
+            ps.setInt(1, key1);
+            ps.setInt(2, key2);
+            ps.setString(3, val1);
+            ps.execute();
+            convertUpsertToMutations(conn);
+        }
+    }
+
+    private void upsertPartialRow(int key1, int key2, int value1)
+            throws SQLException, IOException {
+
+        try(Connection conn = DriverManager.getConnection(getUrl(), new Properties())){
+            PreparedStatement
+                    ps =
+                    conn.prepareStatement(
+                            String.format(partialRowUpsert1, dataTableFullName, FIRST_ID, SECOND_ID,
+                                    SECOND_VALUE));
+            ps.setInt(1, key1);
+            ps.setInt(2, key2);
+            ps.setInt(3, value1);
+            ps.execute();
+            convertUpsertToMutations(conn);
+        }
+    }
+
+    private void upsertCompleteRow(int key1, int key2, String val1
+    , int val2) throws SQLException, IOException {
+        try(Connection conn = DriverManager.getConnection(getUrl(), new Properties())) {
+            PreparedStatement
+                    ps = conn.prepareStatement(String.format(completeRowUpsert, dataTableFullName));
+            ps.setInt(1, key1);
+            ps.setInt(2, key2);
+            ps.setString(3, val1);
+            ps.setInt(4, val2);
+            ps.execute();
+            convertUpsertToMutations(conn);
+        }
+    }
+
+    private void convertUpsertToMutations(Connection conn) throws SQLException, IOException {
+        Iterator<Pair<byte[],List<KeyValue>>>
+                dataTableNameAndMutationKeyValuesIter = PhoenixRuntime.getUncommittedDataIterator(conn);
+        Pair<byte[], List<KeyValue>> elem = dataTableNameAndMutationKeyValuesIter.next();
+        byte[] key = elem.getSecond().get(0).getRow();
+        long mutationTS = EnvironmentEdgeManager.currentTimeMillis();
+
+        for (KeyValue kv : elem.getSecond()) {
+            Cell cell =
+                    CellUtil.createCell(kv.getRow(), kv.getFamily(), kv.getQualifier(),
+                            mutationTS, kv.getType(), kv.getValue());
+            if (KeyValue.Type.codeToType(cell.getTypeByte()) == KeyValue.Type.Put) {
+                if (put == null ) {
+                    put = new Put(key);
+                }
+                put.add(cell);
+            } else {
+                if (delete == null) {
+                    delete = new Delete(key);
+                }
+                delete.addDeleteMarker(cell);
+            }
+        }
+    }
+
+    private void initializeRebuildScannerAttributes() {
+        when(rebuildScanner.setIndexTableTTL(Matchers.anyInt())).thenCallRealMethod();
```

Comment:
All of this partial mocking in initializeRebuildScannerAttributes and initalizeGlobalMockitoSetup should be taken as a "code smell" that the rebuild scanner needs to be broken apart into multiple classes. Generally when mocking you have one class under test which isn't mocked at all, and other dependent classes that aren't being tested that are completely mocked. 

Again, don't have to fix this in _this_ PR, because without the tests first it's hard to refactor this safely. But the refactoring should happen.